src/Storages/S3Queue/StorageS3Queue.cpp

```diff
@@ -0,0 +1,736 @@
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int BAD_ARGUMENTS;
+    extern const int S3_ERROR;
+    extern const int NOT_IMPLEMENTED;
+    extern const int QUERY_NOT_ALLOWED;
+    extern const int NO_ZOOKEEPER;
+    extern const int REPLICA_ALREADY_EXISTS;
+    extern const int INCOMPATIBLE_COLUMNS;
+}
+
+
+StorageS3Queue::StorageS3Queue(
+    std::unique_ptr<S3QueueSettings> s3queue_settings_,
+    const StorageS3::Configuration & configuration_,
+    const StorageID & table_id_,
+    const ColumnsDescription & columns_,
+    const ConstraintsDescription & constraints_,
+    const String & comment,
+    ContextPtr context_,
+    std::optional<FormatSettings> format_settings_,
+    bool distributed_processing_,
+    ASTPtr partition_by_)
+    : IStorage(table_id_)
+    , WithContext(context_)
+    , s3queue_settings(std::move(s3queue_settings_))
+    , s3_configuration{configuration_}
+    , keys({s3_configuration.url.key})
+    , mode(s3queue_settings->mode)
+    , after_processing(s3queue_settings->after_processing)
+    , milliseconds_to_wait(s3queue_settings->s3queue_polling_min_timeout_ms)
+    , format_name(configuration_.format)
+    , compression_method(configuration_.compression_method)
+    , name(s3_configuration.url.storage_name)
+    , distributed_processing(distributed_processing_)
+    , format_settings(format_settings_)
+    , partition_by(partition_by_)
+    , is_key_with_globs(s3_configuration.url.key.find_first_of("*?{") != std::string::npos)
+    , log(&Poco::Logger::get("StorageS3Queue (" + table_id_.table_name + ")"))
+{
+    LOG_INFO(log, "Init engine");
+
+    if (!is_key_with_globs)
+    {
+        throw Exception(ErrorCodes::QUERY_NOT_ALLOWED, "S3Queue engine can read only from url with globs");
+    }
+
+    String setting_zookeeper_path = s3queue_settings->keeper_path;
+    LOG_INFO(log, "Settings zookeeper_path={}", setting_zookeeper_path);
+    if (setting_zookeeper_path.empty())
+    {
+        auto table_id = getStorageID();
+        auto database = DatabaseCatalog::instance().getDatabase(table_id.database_name);
+        bool is_in_replicated_database = database->getEngineName() == "Replicated";
+
+        auto default_path = getContext()->getSettingsRef().s3queue_default_zookeeper_path.value;
+        if (!default_path.empty())
+        {
+            zookeeper_path
+                = zkutil::extractZooKeeperPath(fs::path(default_path) / toString(table_id.uuid), /* check_starts_with_slash */ true, log);
+        }
+        else if (is_in_replicated_database)
+        {
+            LOG_INFO(log, "S3Queue engine keeper_path not specified. Use replicated database zookeeper path");
+            String base_zookeeper_path = assert_cast<const DatabaseReplicated *>(database.get())->getZooKeeperPath();
+            zookeeper_path = zkutil::extractZooKeeperPath(
+                fs::path(base_zookeeper_path) / toString(table_id.uuid), /* check_starts_with_slash */ true, log);
```

Comment:
should we append default_path to the path we got from replicated database or some "/s3queue/"? when reading what is in zookeeper someone might be confused about what lies by this uuid, or someone will also be attempting to use the same path for something else